include/ncengine/ecs/AnyComponent.h

```diff
@@ -11,49 +11,70 @@ class AnyComponent
-         * @param handlers A pointer to the handler for T.
+         * @param obj A valid pointer to the component instance.
+         * @param handlers A valid pointer to the handler for T.
+         * @throw NcError if either the instance or handler are null.
          */
         template<PooledComponent T>
         explicit AnyComponent(T* instance, ComponentHandler<T>* handler)
-            : m_impl{std::make_unique<detail::AnyImplConcrete<T>>(instance, handler)}
-        {
-        }
+            : m_storage{instance, handler} {}
 
         /** @brief Check if this refers to a valid component. */
         explicit operator bool() const noexcept
         {
-            return static_cast<bool>(m_impl);
+            return m_storage.HasValue();
+        }
+
+        /** @brief Check if two AnyComponents refer to the same component. */
+        auto operator==(const AnyComponent& other) const noexcept -> bool
+        {
+            return m_storage == other.m_storage;
+        }
+
+        /** @brief Check if two AnyComponents refer to different components. */
+        auto operator!=(const AnyComponent& other) const noexcept -> bool
+        {
+            return !(*this == other);
         }
 
-        /** @brief Get the value of ComponentHandler::name for the concrete component type. */
+        /**
+         * @brief Get the value of ComponentHandler::name for the concrete component type.
+         * @throw NcError if invoked on a null AnyComponent.
+         */
         auto Name() const -> std::string_view
         {
-            NC_ASSERT(m_impl, "Invalid use of null AnyComponent.");
-            return m_impl->Name();
+            NC_ASSERT(m_storage.HasValue(), "Invalid use of null AnyComponent.");
+            return m_storage.AsImpl()->Name();
         }
 
-        /** @brief Check if ComponentHandler::drawUI is set for the concrete component type. */
+        /**
+         * @brief Check if ComponentHandler::drawUI is set for the concrete component type.
+         * @throw NcError if invoked on a null AnyComponent.
+         */
         auto HasDrawUI() const -> bool
         {
-            NC_ASSERT(m_impl, "Invalid use of null AnyComponent.");
-            return m_impl->HasDrawUI();
+            NC_ASSERT(m_storage.HasValue(), "Invalid use of null AnyComponent.");
+            return m_storage.AsImpl()->HasDrawUI();
         }
 
-        /** @brief Invoke ComponentHandler::drawUI with the component instance */
+        /**
+         * @brief Invoke ComponentHandler::drawUI with the component instance.
+         * @throw NcError if invoked on a null AnyComponent or if HasDrawUI() is false.
+         */
         void DrawUI()
         {
-            NC_ASSERT(m_impl, "Invalid use of null AnyComponent.");
-            m_impl->DrawUI();
+            NC_ASSERT(m_storage.HasValue(), "Invalid use of null AnyComponent.");
+            NC_ASSERT(HasDrawUI(), "No drawUI handler is set for this component type.");
```

Comment:
Would we rather allow calling this if `HasDrawUI() == false` and just do nothing? That's the default behavior, unless a user has set a function, so I think all calling code always has to check `HasDrawUI()` on every item.

Reply:
Per our discussion - be permissive and allow to no-op to get rid of the checks.